File: objc/Project15/Project13/Screenshot.m
//
//  Screenshot.m
//  Project13
//
//  Created by Jinwoo Kim on 1/11/21.
//

#import "Screenshot.h"

@implementation Screenshot
- (instancetype)init {
    self = [super init];
    if (self) {
        self.caption = @"Your text here";
        self.captionFontName = @" HelveticaNeue-Medium";
        self.captionFontSize = 3;
        self.captionColor = NSColor.blackColor;
        self.backgroundImage = @"";
        self.backgroundColorStart = NSColor.clearColor;
        self.backgroundColorEnd = NSColor.clearColor;
        self.dropShadowStrength = 1;
        self.dropShadowTarget = 2;
    }
    
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self.caption = (NSString *)[coder decodeObjectForKey:@"caption"];
    self.captionFontName = (NSString *)[coder decodeObjectForKey:@"captionFontName"];
    self.captionFontSize = [coder decodeIntForKey:@"captionFontSize"];
    self.captionColor = (NSColor *)[coder decodeObjectForKey:@"captionColor"];
    self.backgroundImage = (NSString *)[coder decodeObjectForKey:@"backgroundImage"];
    self.backgroundColorStart = (NSColor *)[coder decodeObjectForKey:@"backgroundColorStart"];
    self.backgroundColorEnd = (NSColor *)[coder decodeObjectForKey:@"backgroundColorEnd"];
    self.dropShadowStrength = [coder decodeIntForKey:@"dropShadowStrength"];
    self.dropShadowTarget = [coder decodeIntForKey:@"dropShadowTarget"];
    
    return self;
}

- (void)endodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:self.caption forKey:@"caption"];
    [coder encodeObject:self.captionFontName forKey:@"captionFontName"];
    [coder encodeInt:(int)self.captionFontSize forKey:@"captionFontSize"];
    [coder encodeObject:self.captionColor forKey:@"captionColor"];
    [coder encodeObject:self.backgroundImage forKey:@"backgroundImage"];
    [coder encodeObject:self.backgroundColorStart forKey:@"backgroundColorStart"];
    [coder encodeObject:self.backgroundColorEnd forKey:@"backgroundColorEnd"];
    [coder encodeInt:(int)self.dropShadowStrength forKey:@"dropShadowStrength"];
    [coder encodeInt:(int)self.dropShadowTarget forKey:@"dropShadowTarget"];
}
@end

File: swift/Project13/Project13/Screenshot.swift
//
//  Screenshot.swift
//  Project13
//
//  Created by Jinwoo Kim on 1/11/21.
//

import Cocoa

enum ScreenshotError: Error {
    case BadData
}

final class Screenshot: NSObject, NSCoding {
    var caption = "Your text here"
    var captionFontName = " HelveticaNeue-Medium"
    var captionFontSize = 3
    var captionColor = NSColor.black
    
    var backgroundImage = ""
    var backgroundColorStart = NSColor.clear
    var backgroundColorEnd = NSColor.clear
    
    var dropShadowStrength = 1
    var dropShadowTarget = 2
    
    
    override init() {
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        caption = aDecoder.decodeObject(forKey: "caption") as! String
        captionFontName = aDecoder.decodeObject(forKey: "captionFontName") as! String
        captionFontSize = aDecoder.decodeInteger(forKey: "captionFontSize")
        captionColor = aDecoder.decodeObject(forKey: "captionColor") as! NSColor
        
        backgroundImage = aDecoder.decodeObject(forKey: "backgroundImage") as! String
        backgroundColorStart = aDecoder.decodeObject(forKey: "backgroundColorStart") as! NSColor
        backgroundColorEnd = aDecoder.decodeObject(forKey: "backgroundColorEnd") as! NSColor
        
        dropShadowStrength = aDecoder.decodeInteger(forKey: "dropShadowStrength")
        dropShadowTarget = aDecoder.decodeInteger(forKey: "dropShadowTarget")
    }
    
    func encode(with aCoder: NSCoder) {
        aCoder.encode(caption, forKey: "caption")
        aCoder.encode(captionFontName, forKey: "captionFontName")
        aCoder.encode(captionFontSize, forKey: "captionFontSize")
        aCoder.encode(captionColor, forKey: "captionColor")

        aCoder.encode(backgroundImage, forKey: "backgroundImage")
        aCoder.encode(backgroundColorStart, forKey: "backgroundColorStart")
        aCoder.encode(backgroundColorEnd, forKey: "backgroundColorEnd")

        aCoder.encode(dropShadowStrength, forKey: "dropShadowStrength")
        aCoder.encode(dropShadowTarget, forKey: "dropShadowTarget")
    }
}
